### PR TITLE
Register file and Get executions functionality

### PR DESCRIPTION
**Pull request by pmahindrakar-oss**
Usage :
1] Returns list of executions in a project domain

```
bin/flytectl get executions  -d development  -p flytesnacks
INFO[0000] Using config file: [config.yaml]
INFO[0000] Config section [adminutils] updated. No update handler registered.  src="viper.go:318"
INFO[0000] Config section [logger] updated. Firing updated event.  src="viper.go:320"
| NAME (4)   | WORKFLOW NAME                                                      | TYPE     | PHASE     | STARTED                        | ELAPSED TIME  |
 ------------ -------------------------------------------------------------------- ---------- ----------- -------------------------------- ---------------
| h5i7q5h9he | recipes.07_juptyer_notebooks.whole_square_workflow.whole_square_wf | WORKFLOW | SUCCEEDED | 2021-01-19T10:42:41.467046Z    | 62.763573800s |
| lqi9q9l8fb | recipes.07_juptyer_notebooks.whole_square_workflow.whole_square_wf | WORKFLOW | ABORTED   | 2021-01-19T10:48:12.546933900Z | 13.067154500s |
| vcfy9487ud | recipes.07_juptyer_notebooks.whole_square_workflow.whole_square_wf | WORKFLOW | ABORTED   | 2021-01-20T04:28:08.272178800Z | 8.317607300s  |
| zgxdehv84b | recipes.07_juptyer_notebooks.whole_square_workflow.whole_square_wf | WORKFLOW | ABORTED   | 2021-01-19T15:34:38.757407800Z | 18.213011300s |
4 rows
```
2] Returns a single execution based on its NAME ($.id.name)

```
bin/flytectl get executions  h5i7q5h9he -d development  -p flytesnacks
INFO[0000] Using config file: [config.yaml]
INFO[0000] Config section [admin] updated. Firing updated event.  src="viper.go:320"
INFO[0000] Config section [adminutils] updated. No update handler registered.  src="viper.go:318"
INFO[0000] Config section [logger] updated. Firing updated event.  src="viper.go:320"
| NAME       | WORKFLOW NAME                                                      | TYPE     | PHASE     | STARTED                     | ELAPSED TIME  |
 ------------ -------------------------------------------------------------------- ---------- ----------- ----------------------------- ---------------
| h5i7q5h9he | recipes.07_juptyer_notebooks.whole_square_workflow.whole_square_wf | WORKFLOW | SUCCEEDED | 2021-01-19T10:42:41.467046Z | 62.763573800s |
1 rows
3] For an invalid name or non existent execution 
```

```
bin/flytectl get executions  WORKFLOW -d development  -p flytesnacks
INFO[0000] Using config file: [config.yaml]
INFO[0000] Config section [logger] updated. Firing updated event.  src="viper.go:320"
Error: rpc error: code = NotFound desc = entry not found
```

4] Register files containing serialized tasks, workflows and launch plans with flyteadmin
```
bin/flytectl register file  _pb_output_new/* -d development  -p flytesnacks -v v52 
INFO[0000] Using config file: [config.yaml]             
INFO[0000] Config section [logger] updated. Firing updated event.  src="viper.go:320"
 ------------------------------------------------------------------------------------------------- --------- ---------------------------------------------------- 
| NAME (9)                                                                                        | STATUS  | ADDITIONAL INFO                                    |
 ------------------------------------------------------------------------------------------------- --------- ---------------------------------------------------- 
| _pb_output_new/000_my-objectstore-sensor_1.pb                                                   | Success | Successfully registered file                       |
 ------------------------------------------------------------------------------------------------- --------- ---------------------------------------------------- 
| _pb_output_new/000_recipes.core.07_juptyer_notebooks.whole_square_workflow.square_1.pb          | Success | Successfully registered file                       |
 ------------------------------------------------------------------------------------------------- --------- ---------------------------------------------------- 
| _pb_output_new/001_recipes.core.07_juptyer_notebooks.whole_square_workflow.sum_1.pb             | Success | Successfully registered file                       |
 ------------------------------------------------------------------------------------------------- --------- ---------------------------------------------------- 
| _pb_output_new/001_recipes.core.advanced.custom_task_plugin.print_file_1.pb                     | Success | Successfully registered file                       |
 ------------------------------------------------------------------------------------------------- --------- ---------------------------------------------------- 
| _pb_output_new/002_recipes.core.07_juptyer_notebooks.whole_square_workflow.whole_square_wf_2.pb | Success | Successfully registered file                       |
 ------------------------------------------------------------------------------------------------- --------- ---------------------------------------------------- 
| _pb_output_new/002_recipes.core.advanced.custom_task_plugin.my_workflow_2.pb                    | Success | Successfully registered file                       |
 ------------------------------------------------------------------------------------------------- --------- ---------------------------------------------------- 
| _pb_output_new/003_recipes.core.07_juptyer_notebooks.whole_square_workflow.whole_square_wf_3.pb | Success | Successfully registered file                       |
 ------------------------------------------------------------------------------------------------- --------- ---------------------------------------------------- 
| _pb_output_new/003_recipes.core.advanced.custom_task_plugin.my_workflow_3.pb                    | Success | Successfully registered file                       |
 ------------------------------------------------------------------------------------------------- --------- ---------------------------------------------------- 
| _pb_output_new/004_my-objectstore-sensor_1.pb                                                   | Failed  | Error registering file due to rpc error: code =    |
|                                                                                                 |         | AlreadyExists desc = identical task already exists |
|                                                                                                 |         | with id resource_type:TASK project:"flytesnacks"   |
|                                                                                                 |         | domain:"development" name:"my-objectstore-sensor"  |
|                                                                                                 |         | version:"v52"                                      |
 ------------------------------------------------------------------------------------------------- --------- ---------------------------------------------------- 
```

5] Re-Register files containing the same serialized tasks,workflows and launchplans
    It throws an error that entity already exists
```
bin/flytectl register file  _pb_output_new/* -d development  -p flytesnacks -v v52 
INFO[0000] Using config file: [config.yaml]             
INFO[0000] Config section [adminutils] updated. No update handler registered.  src="viper.go:318"
INFO[0000] Config section [root] updated. No update handler registered.  src="viper.go:318"
INFO[0000] Config section [logger] updated. Firing updated event.  src="viper.go:320"
 ----------------------------------------------- -------- ---------------------------------------------------- 
| NAME                                          | STATUS | ADDITIONAL INFO                                    |
 ----------------------------------------------- -------- ---------------------------------------------------- 
| _pb_output_new/000_my-objectstore-sensor_1.pb | Failed | Error registering file due to rpc error: code =    |
|                                               |        | AlreadyExists desc = identical task already exists |
|                                               |        | with id resource_type:TASK project:"flytesnacks"   |
|                                               |        | domain:"development" name:"my-objectstore-sensor"  |
|                                               |        | version:"v52"                                      |
 ----------------------------------------------- -------- ---------------------------------------------------- 
```

5] Using skipError
```
bin/flytectl register file  _pb_output_new/* -d development  -p flytesnacks -v v52 --skipOnError
INFO[0000] Using config file: [config.yaml]             
INFO[0000] Config section [logger] updated. Firing updated event.  src="viper.go:320"
 ------------------------------------------------------------------------------------------------- --------- -------------------------------------------------------------------------------- 
| NAME (216)                                                                                      | STATUS  | ADDITIONAL INFO                                                                |
 ------------------------------------------------------------------------------------------------- --------- -------------------------------------------------------------------------------- 
| _pb_output_new/000_my-objectstore-sensor_1.pb                                                   | Failed  | Error registering file due to rpc error: code =                                |
|                                                                                                 |         | AlreadyExists desc = identical task already exists                             |
|                                                                                                 |         | with id resource_type:TASK project:"flytesnacks"                               |
|                                                                                                 |         | domain:"development" name:"my-objectstore-sensor"                              |
|                                                                                                 |         | version:"v52"                                                                  |
 ------------------------------------------------------------------------------------------------- --------- -------------------------------------------------------------------------------- 
| _pb_output_new/000_recipes.core.07_juptyer_notebooks.whole_square_workflow.square_1.pb          | Failed  | Error registering file due to rpc error: code =                                |
|                                                                                                 |         | AlreadyExists desc = identical task already exists                             |
|                                                                                                 |         | with id resource_type:TASK project:"flytesnacks"                               |
|                                                                                                 |         | domain:"development"                                                           |
|                                                                                                 |         | name:"recipes.core.07_juptyer_notebooks.whole_square_workflow.square"          |
|                                                                                                 |         | version:"v52"   
```
## Type
 - [ ] Bug Fix
 - [X] Feature
 - [ ] Plugin

## Are all requirements met?
Pending items :
1] Unit tests for files.go are to be added and would be taken as a followup.
2] Additional options for launch plan registration including auth_role , which are currently available for flytekit and any other additional options to be added.

 - [ ] Code completed
 - [ ] Smoke tested
 - [X] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
Will be linked after resolving the WIP status
https://github.com/lyft/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/lyft/flyte/issues/<number>_

### CODE REVIEW

**Comment by kumare3** [comment]
@pmahindrakar-oss I talked with @wild-endeavor and @katrogan, they made me realize that LaunchPlanSpec does not really have the LaunchPlan Name. This might need the identifier file and/or the name of the file itself.

File: cmd/core/cmd_ctx.go [hunk 1]
```diff
@@ -16,6 +16,10 @@ func (c CommandContext) AdminClient() service.AdminServiceClient {
 	return c.adminClient
 }
 
+func (c *CommandContext) SetAdminClient(adminClient service.AdminServiceClient) {
```
Comment (kumare3) [inline cmd/core/cmd_ctx.go:19]: can we add a constructor here
`NewCommandContext(....)` and use that everywhere

File: cmd/register/config.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package register
+
+import "github.com/lyft/flytestdlib/config"
+
+//go:generate pflags Config
+
+var (
+	defaultConfig = &Config{
```
Comment (kumare3) [inline cmd/register/config.go:8]: default value for version `v1`?

File: cmd/register/register.go [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+		Long: "Takes input files as serialized versions of the tasks/workflows/launchplans and registers them with flyteadmin.\n" +
+			  "Currently these input files are protobuf files generated as output from flytekit serialize.\n" +
+			  "Project & Domain are mandatory fields to be passed for registration and an optional version which defaults to v1\n" +
+			  "If the entities are already registered with flyte for the same version then registration would fail.\n" ,
+	}
+	registerResourcesFuncs := map[string]cmdcore.CommandEntry{
+		"files":    {CmdFunc: registerFromFilesFunc, Aliases: []string{"file"}},
+	}
+	cmdcore.AddCommands(registerCmd, registerResourcesFuncs)
+	registerCmd.PersistentFlags().StringVarP(&(GetConfig().version), "version", "v", "v1", "Specifies the Version to use for registering the serialized files.")
```
Comment (kumare3) [inline cmd/register/register.go:22]: no need if you `go generate`

Comment (pmahindrakar-oss) [reply cmd/register/register.go]: Done

File: cmd/register/config.go (hunk 2 above)
Comment (pmahindrakar-oss) [reply cmd/register/config.go:8]: Done

File: cmd/core/cmd_ctx.go (hunk 1 above)
Comment (pmahindrakar-oss) [reply cmd/core/cmd_ctx.go]: Done.

**Comment by kumare3** [comment]
This `--register.version 12` is what I was afraid of. @EngHabu, so if we use configs for all commands, then we get them as global options. Do you think we should just add them as Flags for the specific sub-command?

File: cmd/register/files.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+		return err
+	default:
+		return errors.New(fmt.Sprintf("Failed registering unknown entity  %v type as part of %v file", message, name))
+	}
+}
+
+func hydrateNode(node *core.Node) {
+	targetNode := node.Target
+	switch targetNode.(type) {
+	case *core.Node_TaskNode:
```
Comment (kumare3) [inline cmd/register/files.go:87]: I think we need to fix this for BranchNode.

Comment (kumare3) [reply cmd/register/files.go:87]: can we add a default case, when none of the case statements match, error out

File: cmd/register/files.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+func hydrateIdentifier(identifier *core.Identifier) {
+	identifier.Project = config.GetConfig().Project
+	identifier.Domain = config.GetConfig().Domain
+	identifier.Version = GetConfig().version
+}
+
+func hydrateSpec(message proto.Message) {
+	switch message.(type) {
+	case *admin.LaunchPlanSpec:
```
Comment (kumare3) [inline cmd/register/files.go:112]: error case

File: cmd/register/files.go [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+		launchSpec := message.(*admin.LaunchPlanSpec)
+		hydrateIdentifier(launchSpec.WorkflowId)
```
Comment (kumare3) [inline cmd/register/files.go:114]: this should now be LaunchPlan

File: cmd/register/files.go [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+		if err != nil {
+			logger.Errorf(ctx, "Error reading file %v  due to : %v. Skipping", absFilePath, err)
+			continue
+		}
+		spec, err := unMarshalContents(ctx, fileContents, absFilePath)
+		if err != nil {
+			logger.Errorf(ctx, "Error unmarshalling Skipping file %v due to : %v", absFilePath, err)
+			continue
+		}
+		name := getEntityNameFromPath(absFilePath)
```
Comment (kumare3) [inline cmd/register/files.go:147]: put a TODO here?

File: cmd/register/files.go (hunk 5 above)
Comment (pmahindrakar-oss) [reply cmd/register/files.go]: Done

File: cmd/register/files.go (hunk 4 above)
Comment (pmahindrakar-oss) [reply cmd/register/files.go:89]: Yes that in the list of pending items. Have added that in this MR itself now.

File: cmd/register/files.go (hunk 6 above)
Comment (pmahindrakar-oss) [reply cmd/register/files.go]: Done . And tested this with flytekit 0.16.0b2

File: cmd/register/files.go (hunk 7 above)
Comment (pmahindrakar-oss) [reply cmd/register/files.go]: Removed this functionality now that we have flytekit 0.16.0b2 release.

File: cmd/register/register.go (hunk 3 above)
Comment (pmahindrakar-oss) [reply cmd/register/register.go:22]: The global flag of --register.version along with other global flags are not getting updated in the Config which is needed by the command actions. I am going back to using persistent flags . Will create an issue to fix this across the board.
Having a persistent flag has an advantage that this flag is only visible under the command context

File: cmd/register/files.go [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+func hydrateIdentifier(identifier *core.Identifier) {
+	if len(identifier.Project) == 0 || identifier.Project == registrationProjectPattern {
+		identifier.Project = config.GetConfig().Project
+	}
+	if len(identifier.Domain) == 0 || identifier.Domain == registrationDomainPattern {
+		identifier.Domain = config.GetConfig().Domain
+	}
+	if len(identifier.Version) == 0 || identifier.Version == registrationVersionPattern {
+		identifier.Version = GetConfig().version
+	}
```
Comment (pmahindrakar-oss) [inline cmd/register/files.go:142]: As per the offline discussion with Ketan , Decision was made to use the registration pattern

**Comment by pmahindrakar-oss** [comment]
> This `--register.version 12` is what I was afraid of. @EngHabu, so if we use configs for all commands, then we get them as global options. Do you think we should just add them as Flags for the specific sub-command?

configAccessor.InitializePflags(rootCmd.PersistentFlags()) 
Causes any place where we call
section = config.MustRegisterSection(key, config)
to register flags mentioned in the config with the prefix key.

Hence for 
section = config.MustRegisterSection("adminutils", defaultConfig)
we see the batchSize, maxRecords part of defaultConfig structure to be available in global flags.
```
Global Flags:
     ......
      --adminutils.batchSize int              Maximum number of records to retrieve per call. (default 100)
      --adminutils.maxRecords int             Maximum number of records to retrieve. (default 500)
    .......
```
Using the PFlagProvider interface approach
```
type PFlagProvider interface {
	GetPFlagSet(prefix string) *pflag.FlagSet
}
```
We are now able to use the flags for specific-subcommand and donot appear in the global flag list.

**Comment by EngHabu** [comment]
Hey @pmahindrakar-oss, Can you clarify what is the end goal and what's the problem? what you described is the expected behavior yes... if you register the config section, the flags will be registered globally (nested ... etc.) if you just use the GetPFlagSet interface method, you can add them to any specific command you want... Is that causing problems?

**Comment by pmahindrakar-oss** [comment]
> Hey @pmahindrakar-oss, Can you clarify what is the end goal and what's the problem? what you described is the expected behavior yes... if you register the config section, the flags will be registered globally (nested ... etc.) if you just use the GetPFlagSet interface method, you can add them to any specific command you want... Is that causing problems?

Hi @EngHabu , The end goal was to have flags at subcommand level and it was unclear if registering the config section allows it. After @kumare3 added the interface for GetPFlagSet in his MR the issue is resolved now and was able to use that approach in the current MR.

**Comment by kumare3** [comment]
@pmahindrakar-oss and @EngHabu  are you guys talking over each other. I am confused

File: cmd/register/files.go [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+}
+
+func hydrateIdentifier(identifier *core.Identifier) {
+	if len(identifier.Project) == 0 || identifier.Project == registrationProjectPattern {
+		identifier.Project = config.GetConfig().Project
+	}
+	if len(identifier.Domain) == 0 || identifier.Domain == registrationDomainPattern {
+		identifier.Domain = config.GetConfig().Domain
+	}
+	if len(identifier.Version) == 0 || identifier.Version == registrationVersionPattern {
```
Comment (kumare3) [inline cmd/register/files.go:154]: for remote entities that use latest version, is the version field optional @katrogan ?

File: cmd/register/files.go [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+			return errors.New(fmt.Sprintf("Unknown type %T", branchNodeWrapper.BranchNode.IfElse.Default))
+		}
+	default:
+		return errors.New(fmt.Sprintf("Unknown type %T", targetNode))
+	}
+	return nil
+}
+
+func hydrateIdentifier(identifier *core.Identifier) {
+	if len(identifier.Project) == 0 || identifier.Project == registrationProjectPattern {
```
Comment (kumare3) [inline cmd/register/files.go:148]:
```suggestion
	if identifier.Project == "" || identifier.Project == registrationProjectPattern {
```

**Comment by kumare3** [comment]
2 small nits, after +1

**Review by kumare3 — APPROVED** [review]

**Pull request merged**